woocommerce/woocommerce
# [Experimental] Product Filters: New and improved blocks structure

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #50871,  closes #50867.

This PR introduces a new block structure as proposed in [https://github.com/woocommerce/woocommerce/issues/50871#issuecomment-2320933584](https://github.com/woocommerce/woocommerce/issues/50871#issuecomment-2320933584):

- Removes the Product Filter block (we keep the code for now, but it will be eventually removed from the source code).
- Updates the Attribute Filter block:
    - Makes it a parent block like the `core/group` block, which contains the Header and the new List block.
    - Moves the rendering logic to the List block. 
    - Passes the filter options down to the List block through context.
- Creates a new List block.
    - Renders the filter options as a list from the data passed through block context.
    - Have advanced settings to customize the list further ~(WIP)~
    - Have improved block select behavior:
        - When users select the block from the canvas by clicking on the filter options preview, the Attribute Filter block is selected.
        - When the Attribute Filter block is selected, users can click on the filter options preview on more time to select the List block.
        - The outline will inform users about the currently selected block.
        - The block select behavior inside the List view are same as other blocks.
- Adds Interactivity API directives to the `Product Filters` block.
- Updates the Product Filters template part, the template now contains only Active and Attribute blocks.
- Note that the Active block hasn't been updated to the new structure yet. 

#### Why should we use the new structure?

For more details, please check #50871. Here is a TL;DR:

- A simple structure that follows the WP Core, where the filter block acts as the `core/query` block, but with some filter settings like attribute selector.
- Avoid hacking the editor by sharing the block settings from the inner block to the parent block.
- Split the responsibility to increase the reusability:
    - The group block (filter blocks) holds the header and filter options. Compared to the `core/group` block, the filter block has some additional responsibilities:
        - Processing the filter data and passing it down through the context for the inner block to consume. 
        - Rendering and updating inner blocks based on the display style setting.
    - The List/Chips/Slider block renders the filter options:
        - Each display style has a corresponding block.
        - The block can be used among multiple filter blocks:
            - List: Attribute, Stock, Rating
            - Chips: Attribute, Active.
        - This paves the way for us and third parties to create more display styles like Color Swatches, Image Swatches, Size Grid, Price List, etc.

![image](https://github.com/user-attachments/assets/8f6e7026-5a05-464e-a241-743f7e79dbe1)

#### Notes:
- The price, stock, and rating filter blocks are removed from the template part as they need updating to remove the wrapper. We will add them back when working on the new design of each block later.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
2. Reset any customization made to the Product Filters template part.
3. Add the template part to the Product Catalog template.
4. Save and check the Shop page on the front end. See:
    1. The Attribute Filter block picks up the correct block to render, with the correct title set to the product attribute label.
    2. Selecting one attribute, see the selected attribute appears in the Active block.
    3. See the Product Collection block updates.
    4. See the URL updates.
5. Edit the Product Catalog template or Product Filters template part:
    1. See the preview renders Active and Attribute filter block.
    2. Open the List view and Sidebar.
    3. From the canvas, click on the filter options. See the Attribute Filter block is selected, not the List block.
    4. Try changing the block setting, see the preview updates.
    6. Click on the filter options again, and see the List block is selected.
    7. Change the color settings of the List block, see the editor preview update as you change.
    8. Chang the display style from List to Chip. See the editor preview change. The Chips block hasn't been worked on, so you will see the placeholder.
    9. Change back to the List block, see the customization you did previously is preserved.
    10. Save the template. Check if it works as before on the frontend.

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
